File: php/2023/03.php
<?php

declare(strict_types=1);

require_once './tools.php';

$handle = fopen('input/03.txt', 'r');

startTimer();
$sum = 0;

$map = [];
while (($line = trim((string) fgets($handle))) !== '') {
    $map[] = str_split($line);
}

// parse the map to do two things :
// - register every possible numbers and the coordinates of their digits
// - register the position of every symbols, and build a list of the coordinates where we could find a number that counts
//
// So that we will be able later to go through all number and the coordinates to check if they count

$currentNumber = '';
/** @var array<string> $currentNumberCoordinates */
$currentNumberCoordinates = [];

$coordinatesPerNumbers = [
    // keys are numbers, as string
    // values are array<array<string>>, since the keys may appear multiple times in the input
];

$valuablesCoordinatesWithNumbers = [
    // keys are the coordinates as string in format x_y. Ie: 1_2.
    // values are the number that we find at this coordinate (for debug)
];

$partNumbersPerStarsCoords = [
    // keys are star coordinates as string
    // value is the number of numbers adjacent
];

foreach ($map as $y => $chars) {
    foreach ($chars as $x => $char) {
        if (is_numeric($char)) {
            $currentNumber .= $char;
            $currentNumberCoordinates[] = "{$x}_$y";

            continue;
        }

        if ($char !== '.') {
            // this is a symbol
            // build list of all its surrounding coordinates where we find a number
            for ($i = $x - 1; $i <= $x + 1 ; $i++) {
                for ($j = $y - 1; $j <= $y + 1 ; $j++) {
                    if ($i === $x && $j === $y) {
                        continue;
                    }

                    $stringCoord = "{$i}_$j";
                    // var_dump("$y $x $j $i " . $map[$j][$i]);
                    if (! isset($valuablesCoordinatesWithNumbers[$stringCoord]) && is_numeric($map[$j][$i])) {
                        $valuablesCoordinatesWithNumbers[$stringCoord] = $map[$j][$i];
                        // var_dump($stringCoord, $x, $y, $char);
                    }
                }
            }
        }

        if ($currentNumber !== '') {
            $coordinatesPerNumbers[$currentNumber] ??= [];
            $coordinatesPerNumbers[$currentNumber][] = $currentNumberCoordinates;

            foreach ($currentNumberCoordinates as $oneDigitCoords) {
                [$digitX, $digitY] = explode('_', $oneDigitCoords);

                // look around to find all the stars
                for ($i = $digitX - 1; $i <= $digitX + 1 ; $i++) {
                    for ($j = $digitY - 1; $j <= $digitY + 1 ; $j++) {
                        if ($i === $digitX && $j === $digitY) {
                            continue;
                        }

                        if (isset($map[$j][$i]) && $map[$j][$i] === '*') {
                            $startCoords = "{$i}_$j";
                            $partNumbersPerStarsCoords[$startCoords] ??= [];

                            if (! in_array($currentNumber, $partNumbersPerStarsCoords[$startCoords], true)) {
                                $partNumbersPerStarsCoords[$startCoords][] = $currentNumber;
                            }
                        }
                    }
                }
            }

            $currentNumber = '';
            $currentNumberCoordinates = [];
        }
    }
}

// now loop on the saved numbers and the coordinates of their digits
// if we found only one that match the saved interesting coordinates, then consider this number
foreach ($coordinatesPerNumbers as $number => $coordSets) {
    foreach ($coordSets as $coordSet) {
        foreach ($coordSet as $stringCoord) {
            if (isset($valuablesCoordinatesWithNumbers[$stringCoord])) {
                $sum += $number;
                continue(2);
            }
        }
    }
}

// var_dump($map, $coordinatesPerNumbers);

printDay("03.1: $sum"); // 7.1 ms (16.5 ms with the added processing for part 2)

// --------------------------------------------------

rewind($handle);
startTimer();
$sum = 0;

foreach ($partNumbersPerStarsCoords as $startCoords => $partNumbers) {
    if (count($partNumbers) !== 2) {
        continue;
    }

    $sum += ($partNumbers[0] * $partNumbers[1]);
}

printDay("03.2: $sum");

File: python/2023/03.py
import json
from traceback import format_list

import utils

# Day 2023.03

f = open("inputs/03.txt", "r")
lines = f.readlines()

# test input
# lines = """467..114..
# ...*......
# ..35..633.
# ......#...
# 617*......
# .....+.58.
# ..592.....
# ......755.
# ...$.*....
# .664.598..""".split("\n")

for i, line in enumerate(lines):
    lines[i] = line.replace("\n", "")


# --------------------------------------------------
# part 1

utils.startTimer()

def checkIfPartNumber(targetX: int, targetY: int) -> bool:
    global lines, numbersPerGear
    maxLineLength: int = len(lines[0])
    maxHeight: int = len(lines)

    for y in range(targetY - 1, targetY + 2):
        if y < 0 or y >= maxHeight:
            continue

        for x in range(targetX - 1, targetX + 2):
            if x < 0 or x >= maxLineLength:
                continue

            if y == targetY and x == targetX:
                continue

            char: str = lines[y][x]
            if char != '.' and not char.isdigit():
                return True

    return False
# end function

currentNumber: str = ''
sum: int = 0
isPartNumber: bool = False

for y, line in enumerate(lines):
    for x, char in enumerate(line):
        if char.isdigit():
            currentNumber += char
            if not isPartNumber:
                isPartNumber = checkIfPartNumber(x, y)
        elif currentNumber != '':
            if isPartNumber:
                sum += int(currentNumber)
                isPartNumber = False
            currentNumber = ''

utils.endPart('1', sum) # 553825


# --------------------------------------------------
# part 2

utils.startTimer()

# keys are the gears coordinates in the for "x_y", values are list of their adjacent numbers
numbersPerGear: dict[str, list[int]] = {}

def checkAdjacentGear(lastDigitX: int, lastDigitY: int, number: int) -> None:
    # note: the coordinates here are the end of the current number
    global lines, numbersPerGear

    maxLineLength: int = len(lines[0])
    maxHeight: int = len(lines)

    # check all digits of the number
    for digitX in range(lastDigitX - len(str(number)) + 1, lastDigitX + 1):
        for y in range(lastDigitY - 1, lastDigitY + 2):
            if y < 0 or y >= maxHeight:
                continue

            for x in range(digitX - 1, digitX + 2):
                if x < 0 or x >= maxLineLength:
                    continue

                if y == lastDigitY and x == digitX:
                    continue

                char: str = lines[y][x]
                if char != '*':
                    continue

                gearCoords: str = f"{x}_{y}"
                if numbersPerGear.get(gearCoords) is None:
                    numbersPerGear[gearCoords] = []

                if number not in numbersPerGear[gearCoords]:
                    numbersPerGear[gearCoords].append(number)
                    return
                    # we make the correct asumption here that no gear is next to two identical number
                    # and that no numbers is next to two gear
# end function

currentNumber: str = ''

for y, line in enumerate(lines):
    if currentNumber != '': # the number ended at the last column
        checkAdjacentGear(len(line) - 1, y - 1, int(currentNumber))
        currentNumber = ''

    for x, char in enumerate(line):
        if char.isdigit():
            currentNumber += char
        elif currentNumber != '':
            checkAdjacentGear(x - 1, y, int(currentNumber))
            currentNumber = ''

sum: int = 0
for gearCoords, numbers in numbersPerGear.items():
    if len(numbers) == 2:
        sum += numbers[0] * numbers[1]

utils.endPart('2', sum) # 93994191
